cpp/include/raft/sparse/linalg/laplacian.cuh

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <raft/core/device_csr_matrix.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/sparse/linalg/detail/laplacian.cuh>
+
+#include <type_traits>
+
+namespace raft {
+namespace sparse {
+namespace linalg {
+
+/** Given a CSR adjacency matrix, return the graph Laplacian
+ *
+ * Note that for non-symmetric matrices, the out-degree Laplacian is returned.
+ */
+template <typename ElementType, typename IndptrType, typename IndicesType, typename NZType>
+auto compute_graph_laplacian(
```

Comment:
Quick comment about `detail`: In RAFT, we follow the pattern where the public header is a shim layer to the detail API. The logic here would be contained in the `detail` namespace as well. For example:

header.cuh:
```
void foo () {
  detail::foo()
}
```

detail/header.cuh:
```
namespace detail {
void foo_kernel() {

}

void foo() {
// all checks and logic
foo_kernel<<< >>> (); 
}

}

Reply:
Yep, that's correct. The publicly facing layer would be responsible for things like input validation (if necessary) but all the guts should be hidden in detail. We've gotten strict about this because we've been bitten too many times by exposing the impl details in the public APIs. RAFT being header-only requires us to be more disciplined and strict with the separation.